source/classes.tex

```diff
@@ -6681,12 +6707,72 @@
 derived-to-base conversions\iref{over.best.ics},
 class member access expressions\iref{expr.ref}, and
 array subscript expressions\iref{expr.sub} applied to \tcode{x}.
-The type of the expression $\tcode{x}_i$ \tcode{<=>} $\tcode{x}_i$
-is denoted by $\tcode{R}_i$.
-It is unspecified
-whether virtual base class subobjects are compared more than once.
+It is unspecified whether virtual base class subobjects
+appear more than once in the expanded list of subobjects.
+
+\rSec2[class.eq]{Equality operators}
+\indextext{operator!equality!defaulted}%
+\indextext{operator!inequality!defaulted}%
+
+\pnum
+A defaulted equality operator\iref{expr.eq} function
+shall have a declared return type \tcode{bool}.
+
+\pnum
+A defaulted \tcode{==} operator function for a class \tcode{C}
+is defined as deleted
+unless, for each $\tcode{x}_i$ in the expanded list of subobjects
+for an object \tcode{x} of type \tcode{C},
+$\tcode{x}_i\tcode{ == }\tcode{x}_i$ is a valid expression and
+contextually convertible to \tcode{bool}.
 
 \pnum
+The return value \tcode{V} of a defaulted \tcode{==} operator function
+with parameters \tcode{x} and \tcode{y} is determined
+by comparing corresponding elements $\tcode{x}_i$ and $\tcode{y}_i$
+in the expanded lists of subobjects for \tcode{x} and \tcode{y}
+until the first index $i$
+where $\tcode{x}_i\tcode{ == }\tcode{y}_i$ yields a result value which,
+when contextually converted to \tcode{bool}, yields \tcode{false}.
+If no such index exists, \tcode{V} is \tcode{true}.
+Otherwise, \tcode{V} is \tcode{false}.
+
+\pnum
+A defaulted \tcode{!=} operator function for a class \tcode{C}
+with parameters \tcode{x} and \tcode{y} is defined as deleted if
+\begin{itemize}
+\item
+  overload resolution\ref{over.match}, as applied to \tcode{x == y}
+  (also considering synthesized candidates
+  with reversed order of parameters\iref{over.match.oper}),
+  results in an ambiguity or a function
+  that is deleted or inaccessible from the operator function, or
+\item
+  \tcode{x == y} or \tcode{y == x}
+  cannot be contextually converted to \tcode{bool}.
+\end{itemize}
+Otherwise, the operator function yields \tcode{(x == y) ? false : true}
+if an operator \tcode{==} with the original order of parameters was selected,
+or \tcode{(y == x) ? false : true} otherwise.
+
+\pnum
+\begin{example}
+\begin{codeblock}
+struct D {
+  int i;
+  friend bool operator==(const D& x, const D& y) = default;     // OK, returns \tcode{x.i == y.i}
+  bool operator!=(const D& z) const = default;  // OK, returns \tcode{(*this == z) ? false : true}
+};
+\end{codeblock}
+\end{example}
+
+\rSec2[class.spaceship]{Three-way comparison}
+\indextext{operator!three-way comparison!defaulted}%
+
+\pnum
+Given an expanded list of subobjects for an object \tcode{x} of type \tcode{C},
+the type of the expression $\tcode{x}_i$ \tcode{<=>} $\tcode{x}_i$
+is denoted by $\tcode{R}_i$.
```

Comment:
Given:
```c++
struct X {};
struct Y {
  X x;
  auto operator<=>(const Y&) = default;
};
```
This tells me that R_0 denotes the type of an ill-formed expression, which is nonsense. I think this should read:
```tex
Given [...],
$\tcode{R}_i$ denotes the type of the expression $\tcode{x}_i$ \tcode{<=>} $\tcode{x}_i$ 
if it is well-formed, and otherwise denotes \tcode{void}.
```

Reply:
I agree, I added the same idea to my wording for [P1186](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1186r1.html#wording) but I like your wording better than what I proposed there (I also changed the <code>R<sub>i</sub></code>'s to <code>S<sub>i</sub></code>'s to avoid confusion with `R` which is totally unrelated - but we can cross that bridge in that paper.)

Reply:
Is this a new bug introduced by this paper, or an existing issue? Seems like an existing issue to me. Is this change even editorial to start with?

Reply:
I agree, this is pre-existing - I hadn't paid enough attention to this wording to notice it before. I'll submit a core issue.

Reply:
Thanks.